### PR TITLE
Azure support

### PR DESCRIPTION
**Pull request by pmuens**
PR to add Microsoft Azure support.

This is a rebased version of the original PR by the Amazing Microsoft OS team (https://github.com/serverless/serverless/pull/1426). We've pulled it in to update it according to the recent changes.

Here's the corresponding issue which outlines what will be changed: https://github.com/serverless/serverless/issues/1493

### CODE REVIEW

File: lib/plugins/azure/deploy/lib/deployFunctions.js [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+      this.deployedFunctions[index].zipFileData = data;
+    });
+
+    return BbPromise.resolve();
+  },
+
+  uploadZipFiles() {
+    this.serverless.cli.log('Uploading zip files to Azure...');
+    const targetWebsite = this.serverless.service.resources.azure.variables.sitename;
+    const user = process.env.AZURE_USERNAME;
```
Comment (christopheranderson) [inline lib/plugins/azure/deploy/lib/deployFunctions.js]: Should be able to get deployment credentials from ARM (.../Microsoft.Web/sites/*/config/publishingcredentials). Asking users to fetch this and set env variables isn't going to be smooth, from experience.

**Comment by flomotlik** [comment]
closed because of V1 branch being removed. Could you reopen against master?

**Comment by mi-hol** [comment]
@pmuens , @flomotlik I wonder if the label 'in-review' is actually correct because it doesn't show activity on the project 'Pull Request Pipeline'. Know a timeline for merge would be very valueable :)

**Comment by pmuens** [comment]
Hey @mi-hol thank you for your comment and the pointers! 👍 

We're working on and investigating multi provider support for the Serverless framework (see this PR, #1858 and #2568) for upcoming releases.

Unfortunately this PR is a little bit outdated since the Serverless Framework moved fast the last couple of weeks / months so it needs some updates so that it's ready to be merged.

**Comment by mi-hol** [comment]
Hi @pmuens, Thanks for explaining
Hi @felixrieseberg , @christopheranderson  what is your perception on this?

**Comment by mi-hol** [comment]
@pmuens it would seem that the original authors either lost interest or are unavailable.
Are you in a position to share plans for supporting Azure?

**Comment by pmuens** [comment]
Hey @mi-hol thanks for getting back! 👍 
Multi provider support is still a very high priority for us. We're currently moving some of those PRs into own plugins (and already finished some) so that our main codebase is clean and every provider is integrated via a plugin.

This PR is a little bit outdated but we'll fix it and migrate it pretty soon!

**Comment by pmuens** [comment]
Azure support will be implemented in a separate plugin by the awesome folks from Microsoft. 🎉 

Here's the link to the repository: https://github.com/serverless/serverless-azure-functions

**Pull request closed without merging**